rosetta/converter/converter.go

```diff
@@ -0,0 +1,120 @@
+// Copyright 2021 Alvalor S.A.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not
+// use this file except in compliance with the License. You may obtain a copy of
+// the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+
+package converter
+
+import (
+	"fmt"
+	"strconv"
+
+	"github.com/onflow/cadence"
+	"github.com/onflow/cadence/encoding/json"
+	"github.com/onflow/flow-go/model/flow"
+
+	"github.com/optakt/flow-dps/models/dps"
+	"github.com/optakt/flow-dps/rosetta/identifier"
+	"github.com/optakt/flow-dps/rosetta/object"
+)
+
+type Converter struct {
+	deposit    flow.EventType
+	withdrawal flow.EventType
+}
+
+func New(gen Generator) (*Converter, error) {
+	deposit, err := gen.TokensDeposited(dps.FlowSymbol)
+	if err != nil {
+		return nil, fmt.Errorf("unable to generate deposit event type: %w", err)
+	}
+	withdrawal, err := gen.TokensWithdrawn(dps.FlowSymbol)
+	if err != nil {
+		return nil, fmt.Errorf("unable to generate withdrawal event type: %w", err)
+	}
+
+	c := Converter{
+		deposit:    flow.EventType(deposit),
+		withdrawal: flow.EventType(withdrawal),
+	}
+
+	return &c, err
+}
+
+func (c *Converter) EventToOperation(event flow.Event) (operation *object.Operation, isRelevant bool, err error) {
```

Comment:
I'm really opposed to named return parameters. In my opinion, they should never have added it to the language. I do understand the attraction of having the parameters being self-explanatory. That being said, I think it's extremely confusing if it's only used rarely.

First, it's not consistent; it creates some processing overhead when you now have to scan for variable declarations in a new place, rather than where you are used to it. This is the only such instance in all of Go, so I find it disruptive.

Second, it's confusing. Suddenly, it's OK to return without any parameters, even if there are return values. You can be faced with a naked `return` and now have to implicitly derive what is being returned from much more context than you would usually need to know.

Overall, I think the additional clarity of the function signature is not worth the confusion and cognitive overhead that the construct adds. It's much clearer to just declare the variables in the function, and use code comments to explain the function signature.

Reply:
Ah I'm glad you are so passionate about the subject, because so am I :p

I despise naked returns, so no worries you will never see even the shadow of one coming from me, but I do see value in named returns, for any function that has non-explicit return values. Whenever I see a function with 3 return values and it's not obvious what it is for me and I need to read the code to know what it returns, that's a problem for me, and that problem is solved by named returns.

While I agree that it's not consistent, I do not think it needs to be. Just like each variable thoughout the project does not need a long name to be explicit/clear, not every function needs named returns, but in some cases I really think that not having them is actually more confusing than having them and not expecting them.

When I feel the need to add named returns, I obviously always prefer to not need them, so in this case an alternative which makes sense to me but did not match the requirements of what you proposed initially would simply be to return a specific error in the case where the event is not relevant, instead of using a boolean. This way, named returns are not needed, and the function is explicit.

I still do prefer having the boolean and named returns here however, because otherwise a consumer of that func might not realize that they are supposed to check for whether or not the error happened because it couldn't be parsed or because it can be ignored.

According to [`A Tour of Go`](https://tour.golang.org/basics/7)

> Go's return values may be named. If so, they are treated as variables defined at the top of the function.
> 
> These names **should be used to document the meaning of the return values**.

Reply:
When I read Go code and I can not immediately follow what is happening, imo something is wrong with the code. In this case, I couldn't make sense of what I was reading because the variables were not declared. It took me quite a bit of time to even realize _why_ I was confused.

The quote say _may_ and _if so_ as well.

Regarding the three return values, I think it's not as uncommon as you think to have a boolean and an error return. I also think the return values are rather self-explanatory; a boolean indicates whether something was returned or not, just like maps and other functions that work the same way, and an error value indicates an error. There is nothing confusing to me about this, and `isRelevant` is imo not actually accurate enough to deserve such a long time. `ok` is just fine, as it always is when you work with a single boolean return.

I can agree with replacing this with two return values and a sentinel error, and I don't think it's a problem that a consumer needs to know that error. This is the case for all sentinel errors, and they are used all the time.

However, when it comes to the named return values, I would like to keep them out of the code base. You say consistency doesn't matter here, but I disagree. Go is so easy to read and understand because they try to only have one way of doing one thing as much as possible, and this clearly violates that approach.